tests/components/google_pubsub/test_init.py

```diff
@@ -0,0 +1,262 @@
+"""The tests for the Google Pub/Sub component."""
+from collections import namedtuple
+from datetime import datetime
+
+import pytest
+
+import homeassistant.components.google_pubsub as google_pubsub
+from homeassistant.components.google_pubsub import DateTimeJSONEncoder as victim
+from homeassistant.const import EVENT_STATE_CHANGED
+from homeassistant.core import split_entity_id
+from homeassistant.setup import async_setup_component
+
+import tests.async_mock as mock
+
+GOOGLE_PUBSUB_PATH = "homeassistant.components.google_pubsub"
+
+
+async def test_datetime():
+    """Test datetime encoding."""
+    time = datetime(2019, 1, 13, 12, 30, 5)
+    assert victim().encode(time) == '"2019-01-13T12:30:05"'
+
+
+async def test_no_datetime():
+    """Test integer encoding."""
+    assert victim().encode(42) == "42"
+
+
+async def test_nested():
+    """Test dictionary encoding."""
+    assert victim().encode({"foo": "bar"}) == '{"foo": "bar"}'
+
+
+@pytest.fixture(autouse=True, name="mock_client")
+def mock_client_fixture():
+    """Mock the pubsub client."""
+    with mock.patch(f"{GOOGLE_PUBSUB_PATH}.pubsub_v1") as client:
+        client.PublisherClient = mock.MagicMock()
+        setattr(
+            client.PublisherClient,
+            "from_service_account_json",
+            mock.MagicMock(return_value=mock.MagicMock()),
+        )
+        yield client
+
+
+@pytest.fixture(autouse=True, name="mock_os")
+def mock_os_fixture():
+    """Mock the OS cli."""
+    with mock.patch(f"{GOOGLE_PUBSUB_PATH}.os") as os_cli:
+        os_cli.path = mock.MagicMock()
+        setattr(os_cli.path, "join", mock.MagicMock(return_value="path"))
+        yield os_cli
+
+
+@pytest.fixture(autouse=True)
+def mock_bus_and_json(hass, monkeypatch):
+    """Mock the event bus listener and os component."""
+    hass.bus.listen = mock.MagicMock()
+    monkeypatch.setattr(
+        f"{GOOGLE_PUBSUB_PATH}.json.dumps", mock.Mock(return_value=mock.MagicMock())
+    )
+
+
+async def test_minimal_config(hass, mock_client):
+    """Test the minimal config and defaults of component."""
+    config = {
+        google_pubsub.DOMAIN: {
+            "project_id": "proj",
+            "topic_name": "topic",
+            "credentials_json": "creds",
+            "filter": {},
+        }
+    }
+    assert await async_setup_component(hass, google_pubsub.DOMAIN, config)
+    await hass.async_block_till_done()
+    assert hass.bus.listen.called
+    assert EVENT_STATE_CHANGED == hass.bus.listen.call_args_list[0][0][0]
+    assert mock_client.PublisherClient.from_service_account_json.call_count == 1
+    assert (
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5745133)